domain/machine/state/state.go

```diff
@@ -776,3 +777,78 @@ func (st *State) GetMachineUUID(ctx context.Context, name machine.Name) (string,
 	})
 	return uuid.UUID, errors.Annotatef(err, "getting UUID for machine %q", name)
 }
+
+// KeepInstance reports whether a machine, when removed from Juju, should cause
+// the corresponding cloud instance to be stopped.
+func (st *State) KeepInstance(ctx context.Context, mName machine.Name) (bool, error) {
+	db, err := st.DB()
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+
+	machineNameParam := machineName{Name: mName}
+	result := keepInstance{}
+	query := `
+SELECT &keepInstance.keep_instance 
+FROM   machine 
+WHERE  name = $machineName.name`
+	queryStmt, err := st.Prepare(query, machineNameParam, result)
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err := tx.Query(ctx, queryStmt, machineNameParam).Get(&result)
+		if errors.Is(err, sqlair.ErrNoRows) {
+			return machineerrors.MachineNotFound
+		}
+		if err != nil {
+			return fmt.Errorf("querying machine %q keep instance: %w", mName, err)
+		}
+		return nil
+	})
+	if err != nil {
+		return false, fmt.Errorf("check for machine %q keep instane: %w", mName, err)
+	}
+
+	return result.KeepInstance, nil
+}
+
+// SetKeepInstance sets whether the machine cloud instance will be retained
+// when the machine is removed from Juju. This is only relevant if an instance
+// exists.
+func (st *State) SetKeepInstance(ctx context.Context, mName machine.Name, keep bool) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// Prepare query for updating machine keep instance.
+	machineNameParam := machineName{Name: mName}
+	keepInstanceParam := keepInstance{KeepInstance: keep}
+	query := `
+UPDATE machine 
+SET    keep_instance = $keepInstance.keep_instance 
+WHERE  name = $machineName.name`
+	queryStmt, err := st.Prepare(query, keepInstanceParam, machineNameParam)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	return db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		// Update machine keep instance.
+		var outcome sqlair.Outcome
+		err = tx.Query(ctx, queryStmt, keepInstanceParam, machineNameParam).Get(&outcome)
+		affected, err := outcome.Result().RowsAffected()
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if affected != 1 {
+			return machineerrors.MachineNotFound
+		}
+		if err != nil {
+			return fmt.Errorf("setting keep instance for machine %q: %w", mName, err)
+		}
```

Comment:
I'd rather we split this into two queries, one to get the machine and then one to update. We're then not attempting to write if the machine doesn't exist.

Reply:
Yeah, nice catch. See 7699e5140241109e1fa299cb575ebf398c6fab75